gui/cp437-table.lua

```diff
@@ -1,118 +1,95 @@
+function CPDialog:render_board(dc)
+    for ch = 0,255 do
         if dfhack.screen.charToKey(ch) then
-            p:seek(1 + (ch % 32), 3 + math.floor(ch / 32)):char(ch)
+            dc:seek(ch % 32, math.floor(ch / 32)):char(ch)
         end
     end
-    p:seek(1, self.frame_height - 3):string('Mouse to select')
-    p:seek(1, self.frame_height - 2):key('LEAVESCREEN'):string(': Cancel, ')
-     :key('SELECT'):string(': Done')
 end
 
-function ctable:insert(ch)
-    if type(ch) == 'number' then ch = string.char(ch) end
-    if ch == '\0' then
-        if self.cursor > 1 then
-            self.entry = self.entry:sub(1, self.cursor - 2) .. self.entry:sub(self.cursor)
-            self.cursor = math.max(self.cursor - 1, 1)
-        end
-    else
-        self.entry = self.entry:sub(1, self.cursor - 1) .. ch .. self.entry:sub(self.cursor)
-        self.cursor = self.cursor + 1
-    end
-end
-
-function ctable:simulate()
+function CPDialog:submit()
     local keys = {}
-    for i = 1, #self.entry do
-        local k = dfhack.screen.charToKey(string.byte(self.entry:sub(i, i)))
+    local text = self.subviews.edit.text
+    for i = 1,#text do
+        local k = dfhack.screen.charToKey(string.byte(text:sub(i, i)))
         if not k then
-            qerror(('Invalid character at position %d: %s'):format(i, self.entry:sub(i, i)))
+            dialog.showMessage('Error',
+                ('Invalid character at position %d: "%s"'):
+                    format(i, text:sub(i, i)),
+                COLOR_LIGHTRED)
+            return
         end
         keys[i] = k
     end
     for i, k in pairs(keys) do
         gui.simulateInput(self._native.parent, k)
     end
-    return true
+    self:dismiss()
 end
 
-function ctable:onInput(keys)
-    if keys.LEAVESCREEN then
-        self:dismiss()
-    elseif keys.SELECT then
-        local ret, err = dfhack.pcall(function() self:simulate() end)
-        if ret then
-            self:dismiss()
-        else
-            dialog.showMessage('Error', err.message, COLOR_LIGHTRED)
-        end
-    elseif keys._MOUSE_L_DOWN then
-        local gx = gps.mouse_x - self.frame_rect.x1 - 2
-        local gy = gps.mouse_y - self.frame_rect.y1 - 4
-        if gx >= 0 and gx <= 31 and gy >= 0 and gy <= 7 then
-            local ch = gx + (32 * gy)
-            if ch ~= 0 and dfhack.screen.charToKey(ch) then
-                self:insert(ch)
-            end
-        end
-    elseif keys._STRING then
-        for k in pairs(keys) do
-            if df.interface_key[k] then
-                ch = dfhack.screen.keyToChar(df.interface_key[k])
-                if ch then
-                    self:insert(ch)
-                end
-            end
+function CPDialog:onInput(keys)
+    local x, y = self.subviews.board:getMousePos()
+    if keys._MOUSE_L_DOWN and x then
+        local ch = x + (32 * y)
+        if ch ~= 0 and dfhack.screen.charToKey(ch) then
+            self.subviews.edit:insert(string.char(ch))
```

Comment:
Where's this method defined? EditField doesn't have an insert() method on develop

Reply:
oops, I may not have made a PR for that yet. One second.

Reply:
DFHack/dfhack#2551